src/insn_builder.rs

```diff
@@ -0,0 +1,1389 @@
+// Copyright 2017 <alex.dukhno@icloud.com>
+//
+// Licensed under the Apache License, Version 2.0 <http://www.apache.org/licenses/LICENSE-2.0> or
+// the MIT license <http://opensource.org/licenses/MIT>, at your option. This file may not be
+// copied, modified, or distributed except according to those terms.
+
+//! Module provides API to create eBPF programs by Rust programming language
+
+use ebpf::*;
+
+/// Represents single eBPF instruction
+pub trait Instruction: Sized {
+    /// returns instruction opt code
+    fn opt_code_byte(&self) -> u8;
+
+    /// returns destination register
+    fn get_dst(&self) -> u8 {
+        self.get_insn().dst
+    }
+
+    /// returns source register
+    fn get_src(&self) -> u8 {
+        self.get_insn().src
+    }
+
+    /// returns offset bytes
+    fn get_off(&self) -> i16 {
+        self.get_insn().off
+    }
+
+    /// returns immediate value
+    fn get_imm(&self) -> i32 {
+        self.get_insn().imm
+    }
+
+    /// sets destination register
+    fn set_dst(mut self, dst: u8) -> Self {
+        self.get_insn_mut().dst = dst;
+        self
+    }
+
+    /// sets source register
+    fn set_src(mut self, src: u8) -> Self {
+        self.get_insn_mut().src = src;
+        self
+    }
+
+    /// sets offset bytes
+    fn set_off(mut self, offset: i16) -> Self {
+        self.get_insn_mut().off = offset;
+        self
+    }
+
+    /// sets immediate value
+    fn set_imm(mut self, imm: i32) -> Self {
+        self.get_insn_mut().imm = imm;
+        self
+    }
+
+    fn get_insn(&self) -> &Insn;
+
+    fn get_insn_mut(&mut self) -> &mut Insn;
+}
+
+pub trait IntoBytes {
+    type Bytes;
+
+    fn into_bytes(self) -> Self::Bytes;
+}
+
+impl<'i, I: Instruction> IntoBytes for &'i I {
+    type Bytes = Vec<u8>;
+
+    fn into_bytes(self) -> Self::Bytes {
+        let mut buffer = Vec::with_capacity(8);
+        buffer.push(self.opt_code_byte());
+        buffer.push(self.get_src() << 4 | self.get_dst());
+        buffer.push((self.get_off() & 0x00_ff) as u8);
+        buffer.push(((self.get_off() & 0xff_00) >> 8) as u8);
+        buffer.push((self.get_imm() & 0x00_00_00_ff) as u8);
+        buffer.push(((self.get_imm() & 0x00_00_ff_00) >> 8) as u8);
+        buffer.push(((self.get_imm() & 0x00_ff_00_00) >> 16) as u8);
+        buffer.push(((self.get_imm() & 0xff_00_00_00) >> 24) as u8);
+        buffer
+    }
+}
```

Comment:
Two warnings are raised by this function, regarding the int values that are used for the masks on lines 79 and 83. In fact, all those masks are useless, and you can do as follows:

```
    fn into_bytes(self) -> Self::Bytes {
        let mut buffer = Vec::with_capacity(8);
        buffer.push( self.opt_code_byte());
        buffer.push( self.get_src() << 4 | self.get_dst());
        buffer.push( self.get_off()        as u8);
        buffer.push((self.get_off() >> 8)  as u8);
        buffer.push( self.get_imm()        as u8);
        buffer.push((self.get_imm() >> 8)  as u8);
        buffer.push((self.get_imm() >> 16) as u8);
        buffer.push((self.get_imm() >> 24) as u8);
        buffer
    }
```

Casting to `u8` after the division automatically trims the unwanted part on the “left”.

(Of course you don't have to keep the same formating, aligning stuff is just my personal preference in such case).